src/wallet/wallet.cpp

```diff
@@ -82,15 +82,59 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
-// Custom deleter for shared_ptr<CWallet>.
-static void ReleaseWallet(CWallet* wallet)
+static Mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cv;
+static std::set<CWallet*> g_unloading_wallet_set;
+
+static void DeleteWallet(CWallet* wallet)
 {
     wallet->WalletLogPrintf("Releasing wallet\n");
     wallet->BlockUntilSyncedToCurrentChain();
     wallet->Flush();
+    UnregisterValidationInterface(wallet);
     delete wallet;
 }
 
+// Custom deleter for shared_ptr<CWallet>.
+static void ReleaseWallet(CWallet* wallet)
+{
+    bool unloading{false};
+    {
+        LOCK(g_wallet_release_mutex);
+        unloading = g_unloading_wallet_set.erase(wallet) == 1;
+    }
+    if (unloading) {
+        // Delegate wallet deletion for UnloadWallet caller.
+        g_wallet_release_cv.notify_all();
+    } else {
+        // Delegate wallet deletion for UnloadWallet caller.
```

Comment:
Could say it's not necessary to call `NotifyUnload` because at this point there are no shared pointers to this wallet.